File: Blink_Led/Blink_Led_Gpiod.cpp
#include <gpiod.hpp>
#include <iostream>
#include <chrono>
#include <thread>

int main() {
    // In v2, we use the path to the device
    std::string chip_path = "/dev/gpiochip0"; 
    unsigned int offset = 23; 

    try {
        gpiod::chip  chip = gpiod::chip(chip_path);
        gpiod::line_request request = chip.prepare_request()
            .set_consumer("BlinkApp")
            .add_line_settings(
                offset,
                gpiod::line_settings()
                    .set_direction(gpiod::line::direction::OUTPUT)
            )
            .do_request();

        std::cout << "Blinking LED on offset " << offset << "..." << std::endl;

        while (true) {
            // Toggle logic in v2
            auto current_val = request.get_value(offset);
            auto new_val = (current_val == gpiod::line::value::ACTIVE) ? 
                            gpiod::line::value::INACTIVE : gpiod::line::value::ACTIVE;
            
            request.set_value(offset, new_val);
            
            std::this_thread::sleep_for(std::chrono::milliseconds(500));
        }
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        return 1;
    }
    return 0;
}

File: Blink_Led/python/blink_led_gpiod.py
import gpiod
import time
from gpiod.line import Direction, Value

# Define our chip and pin (offset)
CHIP_PATH = "/dev/gpiochip0"
LINE_OFFSET = 23

# 1. Open the chip and request the line as output
# The 'with' statement automatically closes the chip when finished
with gpiod.request_lines(
    CHIP_PATH,
    consumer="BlinkPy",
    config={
        LINE_OFFSET: gpiod.LineSettings(
            direction=Direction.OUTPUT,
            output_value=Value.INACTIVE
        )
    }
) as request:
    print(f"Blinking LED on {CHIP_PATH} offset {LINE_OFFSET}...")
    
    try:
        while True:
            # 2. Get current value and flip it
            current_val = request.get_value(LINE_OFFSET)
            new_val = Value.ACTIVE if current_val == Value.INACTIVE else Value.INACTIVE
            
            request.set_value(LINE_OFFSET, new_val)
            
            print(f"LED is {'ON' if new_val == Value.ACTIVE else 'OFF'}")
            time.sleep(0.5)
            
    except KeyboardInterrupt:
        print("\nStopping...")
        # Pins are automatically released here